features/netsocket/emac-drivers/TARGET_NXP_EMAC/TARGET_LPC546XX/lpc546xx_emac.cpp

```diff
@@ -0,0 +1,601 @@
+
+    while ((!link) && timeout) {
+        PHY_GetLinkStatus(ENET, phyAddr, &link);
+        if (link) {
+            /* Get link information from PHY */
+            PHY_GetLinkSpeedDuplex(ENET, phyAddr, &phy_speed, &phy_duplex);
+
+            /* Change the MII speed and duplex for actual link status. */
+            config.miiSpeed = (enet_mii_speed_t)phy_speed;
+            config.miiDuplex = (enet_mii_duplex_t)phy_duplex;
+        }
+        timeout--;
+    }
+
+    if (!link) {
+        return false;
+    }
+
+    AT_NONCACHEABLE_SECTION_ALIGN(static enet_rx_bd_struct_t rx_desc_start_addr[ENET_RX_RING_LEN], ENET_BUFF_ALIGNMENT);
+    AT_NONCACHEABLE_SECTION_ALIGN(static enet_tx_bd_struct_t tx_desc_start_addr[ENET_TX_RING_LEN], ENET_BUFF_ALIGNMENT);
+
+    /* Create buffers for each receive BD */
+    for (i = 0; i < ENET_RX_RING_LEN; i++) {
+        rx_buff[i] = memory_manager->alloc_heap(ENET_ETH_MAX_FLEN, ENET_BUFF_ALIGNMENT);
+        if (NULL == rx_buff[i])
+            return false;
+
+        rx_ptr[i] = (uint32_t)memory_manager->get_ptr(rx_buff[i]);
+    }
+
+     /* prepare the buffer configuration. */
+    enet_buffer_config_t buffCfg = {
+        ENET_RX_RING_LEN,
+        ENET_TX_RING_LEN,
+        &tx_desc_start_addr[0],
+        &tx_desc_start_addr[0],
+        &rx_desc_start_addr[0],
+        &rx_desc_start_addr[ENET_RX_RING_LEN],
+        rx_ptr,
+        ENET_ALIGN(ENET_ETH_MAX_FLEN, ENET_BUFF_ALIGNMENT),
+    };
+
+    ENET_Init(ENET, &config, hwaddr, refClock);
+
+    /* Enable the tx & rx interrupt. */
+    ENET_EnableInterrupts(ENET, kENET_DmaTx | kENET_DmaRx);
+
+    /* Create the handler. */
+    ENET_CreateHandler(ENET, &g_handle, &config, &buffCfg, &LPC546XX_EMAC::ethernet_callback, this);
+
+    /* Initialize Descriptor. */
+    ENET_DescriptorInit(ENET, &config, &buffCfg);
+
+    return true;
+}
+
+/** \brief  Allocates a emac_mem_buf_t and returns the data from the incoming packet.
+ *
+ *  \return a emac_mem_buf_t filled with the received packet (including MAC header)
+ *         NULL on memory error
+ */
+emac_mem_buf_t *LPC546XX_EMAC::low_level_input()
+{
+    enet_rx_bd_ring_t *rxBdRing = (enet_rx_bd_ring_t *)&g_handle.rxBdRing[0];
+    enet_rx_bd_struct_t *bdPtr = rxBdRing->rxBdBase + rxBdRing->rxGenIdx;
+    emac_mem_buf_t *p = NULL;
+    emac_mem_buf_t *temp_rxbuf = NULL;
+    uint32_t length = 0;
+
+    /* Determine if a frame has been received */
+    if ((bdPtr->control & ENET_RXDESCRIP_WR_OWN_MASK) != 0) {
+        return p;
+    }
+
+#ifdef LOCK_RX_THREAD
+    /* Get exclusive access */
+    TXLockMutex.lock();
+#endif
+
+    /* Determine if the received frame has an error */
+    if ((bdPtr->control & ENET_RXDESCRIP_WR_ERRSUM_MASK) != 0) {
+        /* Re-use the same buffer in case of error */
+        update_read_buffer(bdPtr, NULL);
+    } else {
+        if (bdPtr->control & ENET_RXDESCRIP_WR_LD_MASK) {
+            length = (bdPtr->control & ENET_RXDESCRIP_WR_PACKETLEN_MASK);
```

Comment:
When testing, we are seeing here that the length here is 4 bytes more than expected. Also with incoming message with 1500 bytes long payload there is a buffer tail overwrite of the received emac_mem_buf_t. Could you check the length handling.

Reply:
I can't see an issue with this code. I need to debug more why this is happening.

Reply:
The datasheet says that the packet written to the buffer includes the CRC, and the length in this descriptor reflects that.

You just need to subtract 4, as far as I can see, as we want the packet length including MAC header, but excluding CRC. (You or the hardware are assumed to have checked and stripped the CRC - the hardware here has checked it but not stripped it.)